packages/@aws-cdk/aws-servicecatalog/test/product.test.ts

```diff
@@ -189,6 +191,66 @@ describe('Product', () => {
     expect(assembly.stacks[0].assets.length).toBe(1);
   }),
 
+  test('product test from product stack with versioning strategy retain', () => {
+    const productStack = new servicecatalog.ProductStack(stack, 'ProductStack');
+
+    new sns.Topic(productStack, 'SNSTopicProductStack');
+
+    new servicecatalog.CloudFormationProduct(stack, 'MyProduct', {
+      productName: 'testProduct',
+      owner: 'testOwner',
+      productVersions: [
+        {
+          productVersionName: 'v1',
+          cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(productStack),
+          versioningStrategy: VersioningStrategy.RETAIN_PREVIOUS_VERSIONS,
+        },
+      ],
+    });
+
+    const assembly = app.synth();
+    expect(assembly.artifacts.length).toEqual(2);
+    expect(assembly.stacks[0].assets.length).toBe(1);
+    expect(assembly.stacks[0].assets[0].path).toEqual('ProductStack.product.template.json');
+
+    const expectedTemplateFileKey = 'MyProduct.ProductStack.v1.product.template.json';
+    const fileExistsContext = fs.existsSync(path.join(PRODUCT_STACK_CONTEXT_DIRECTORY, expectedTemplateFileKey));
+    expect(fileExistsContext).toBe(true);
+  }),
+
+  test('product test from product stack context', () => {
+    new servicecatalog.CloudFormationProduct(stack, 'MyProduct', {
+      productName: 'testProduct',
+      owner: 'testOwner',
+      productVersions: [
+        {
+          productVersionName: 'v1',
+          cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStackContext('ProductStack'),
+        },
+      ],
+    });
+
+    const assembly = app.synth();
+    expect(assembly.artifacts.length).toEqual(2);
+    expect(assembly.stacks[0].assets.length).toBe(1);
+    expect(assembly.stacks[0].assets[0].path).toEqual('asset.c0d25a2f0dec85d4ebb373daf067bb05f6c35c669e6dddf499e1931d794d547f.json');
+  }),
+
+  test('fails product from product stack context not found', () => {
+    expect(() => {
+      new servicecatalog.CloudFormationProduct(stack, 'MyProduct', {
+        productName: 'testProduct',
+        owner: 'testOwner',
+        productVersions: [
+          {
+            productVersionName: 'v2',
+            cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStackContext('ProductStack'),
+          },
+        ],
+      });
+    }).toThrowError('Template MyProduct.ProductStack.v2.product.template.json cannot be found in product stack context');
+  }),
+
```

Comment:
I feel like we're missing a test here for when `versioningStrategy` is `RETAIN`, but the `ProductStack` code chages between the time when the template was saved on disk, and then synthesized again? (That should result in an exception being thrown)